File: zombicideandroid/src/main/java/cc/game/zombicide/android/ZServerMgr.java
package cc.game.zombicide.android;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import cc.lib.logger.Logger;
import cc.lib.logger.LoggerFactory;
import cc.lib.net.ClientConnection;
import cc.lib.net.GameCommand;
import cc.lib.net.GameServer;
import cc.lib.zombicide.ZPlayerName;
import cc.lib.zombicide.ZUser;
import cc.lib.zombicide.p2p.ZUserMP;
import cc.lib.zombicide.ui.UIZombicide;

/**
 * Created by Chris Caron on 7/28/21.
 */
class ZServerMgr extends ZMPCommon implements GameServer.Listener, ZMPCommon.SVR {

    static Logger log = LoggerFactory.getLogger(ZServerMgr.class);

    final GameServer server;

    int colorAssigner = 0;
    final int maxCharacters; // max chars each player can be assigned

    Map<ZPlayerName, Assignee> playerAssignments = new LinkedHashMap<>();
    Map<ClientConnection, ZUser> clientToUserMap = new ConcurrentHashMap<>();
    final CharacterChooserDialog playerChooser;

    ZServerMgr(ZombicideActivity activity, UIZombicide game, int maxCharacters, GameServer server) {
        super(activity, game);
        this.server = server;
        server.addListener(this);
        this.maxCharacters = maxCharacters;
        playerAssignments.clear();
        for (ZombicideActivity.CharLock c : activity.charLocks) {
            Assignee a = new Assignee(c);
            playerAssignments.put(c.player, a);
        }
        activity.user.setColor(nextColor());
        List<Assignee> assignments = new ArrayList<>(playerAssignments.values());
        Collections.sort(assignments);
        playerChooser = new CharacterChooserDialog(activity, assignments, maxCharacters) {
            @Override
            protected void onAssigneeChecked(Assignee assignee) {
                synchronized (playerAssignments) {
                    if (assignee.checked) {
                        assignee.userName = activity.getDisplayName();
                        assignee.color = activity.user.getColorId();
                        assignee.isAssingedToMe = true;
                        game.addCharacter(assignee.name);
                        activity.user.addCharacter(assignee.name);
                    } else {
                        assignee.userName = null;
                        assignee.color = -1;
                        assignee.isAssingedToMe = false;
                        activity.user.removeCharacter(assignee.name);
                        game.removeCharacter(assignee.name);
                    }
                    postNotifyUpdateAssignee(assignee);
                    server.broadcastCommand(newAssignPlayer(assignee));
                    game.boardRenderer.redraw();
                }
            }

            @Override
            protected void onStart() {
                game.setServer(server);
                activity.startGame();
            }
        };
    }

    int nextColor() {
        int color = colorAssigner;
        colorAssigner = (colorAssigner + 1) % ZUser.USER_COLORS.length;
        return color;
    }

    @Override
    public void onConnected(ClientConnection conn) {
        // see if there is an existing user we can reuse
        for (ClientConnection c : clientToUserMap.keySet()) {
            if (!c.isConnected()) {
                ZUser user = clientToUserMap.get(c);
                if (user != null && user.getCharacters().size() > 0) {
                    // reuse
                    clientToUserMap.remove(c);
                    clientToUserMap.put(conn, user);
                    game.addUser(user);
                    user.setName(conn.getDisplayName());
                    conn.sendCommand(newLoadQuest(game.getQuest().getQuest()));
                    conn.sendCommand(newInit(user.getColorId(), maxCharacters, new ArrayList<>(playerAssignments.values())));
                    game.characterRenderer.addMessage(conn.getDisplayName() + " has joined", user.getColor());
                    user.setCharactersHidden(false);
                    game.boardRenderer.redraw();
                    return;
                }
            }
        }
        ZUser user = new ZUserMP(conn);
        int color = nextColor();
        user.setColor(color);
        clientToUserMap.put(conn, user);
        conn.sendCommand(newLoadQuest(game.getQuest().getQuest()));
        conn.sendCommand(newInit(color, maxCharacters, new ArrayList<>(playerAssignments.values())));
        game.characterRenderer.addMessage(conn.getDisplayName() + " has joined", user.getColor());
    }

    @Override
    public void onReconnection(ClientConnection conn) {
        ZUser user = clientToUserMap.get(conn);
        if (user != null) {
            game.addUser(user);
            conn.sendCommand(newLoadQuest(game.getQuest().getQuest()));
            conn.sendCommand(newInit(user.getColorId(), maxCharacters, new ArrayList<>(playerAssignments.values())));
            game.characterRenderer.addMessage(conn.getDisplayName() + " has rejoined", user.getColor());
            user.setCharactersHidden(false);
            game.boardRenderer.redraw();
        }
    }

    @Override
    public void onClientDisconnected(ClientConnection conn) {
        // TODO: Put up a dialog waiting for client to reconnect otherwise set their charaters to invisible and to stop moving
        ZUser user = clientToUserMap.get(conn);
        if (user != null) {
            user.setCharactersHidden(true);
            game.removeUser(user);
            game.characterRenderer.addMessage(conn.getDisplayName() + " has disconnected", user.getColor());
            game.boardRenderer.redraw();
        }
    }

    @Override
    public void onCommand(ClientConnection conn, GameCommand cmd) {
        parseCLCommand(conn, cmd);
    }

    @Override
    public void onChooseCharacter(ClientConnection conn, ZPlayerName name, boolean checked) {
        synchronized (playerAssignments) {
            Assignee a = playerAssignments.get(name);
            ZUser user = clientToUserMap.get(conn);
            if (a != null && user != null) {
                a.checked = checked;
                a.isAssingedToMe = false;
                if (checked) {
                    a.userName = conn.getDisplayName();
                    a.color = user.getColorId();
                    game.addCharacter(name);
                    user.addCharacter(name);
                } else {
                    a.color = -1;
                    a.userName = null;
                    game.removeCharacter(name);
                    user.removeCharacter(name);
                }
                server.broadcastCommand(newAssignPlayer(a));
                playerChooser.postNotifyUpdateAssignee(a);
                game.boardRenderer.redraw();
            }
        }
    }

    @Override
    public void onStartPressed(ClientConnection conn) {
        ZUser user = clientToUserMap.get(conn);
        if (user != null) {
            game.addUser(user);
            conn.sendCommand(newUpdateGameCommand(game));
        }
    }

    @Override
    public void onUndoPressed(ClientConnection conn) {
        activity.runOnUiThread(() -> game.undo());
    }

    @Override
    public void onError(Exception e) {
        log.error(e);
        game.addPlayerComponentMessage("Error:" + e.getClass().getSimpleName() + ":" + e.getMessage());
    }

    @Override
    public void onClientHandleChanged(ClientConnection conn, String newHandle) {
        ZUser user = clientToUserMap.get(conn);
        if (user != null) {
            user.setName(newHandle);
        }
    }
}

File: zombicideandroid/src/main/java/cc/game/zombicide/android/ZServerMgr.kt
package cc.game.zombicide.android

import cc.game.zombicide.android.ZMPCommon.SVR
import cc.lib.logger.LoggerFactory
import cc.lib.net.AClientConnection
import cc.lib.net.AGameServer
import cc.lib.net.ConnectionStatus
import cc.lib.net.GameCommand
import cc.lib.utils.rotate
import cc.lib.utils.takeIfInstance
import cc.lib.zombicide.ZPlayerName
import cc.lib.zombicide.ZUser
import cc.lib.zombicide.p2p.ZUserMP
import cc.lib.zombicide.ui.ConnectedUser
import cc.lib.zombicide.ui.UIZombicide
import java.util.Collections
import java.util.concurrent.ConcurrentHashMap

/**
 * Created by Chris Caron on 7/28/21.
 */
class ZServerMgr(activity: ZombicideActivity, game: UIZombicide, val maxCharacters: Int, val server: AGameServer)
	: SVR(activity, game), AGameServer.Listener, AClientConnection.Listener, ZMPCommon.SVRListener {
	var colorAssigner = 0
	var numStarted = 0
	var playerAssignments: MutableMap<ZPlayerName, Assignee> = Collections.synchronizedMap(LinkedHashMap())
	var clientToUserMap: MutableMap<String, ZUser> = ConcurrentHashMap()
	var playerChooser: CharacterChooserDialogMP? = null
	fun nextColor(): Int {
		val color = colorAssigner
		colorAssigner = colorAssigner.rotate(ZUser.USER_COLORS.size)
		return color
	}

	fun shutdown() {
		server.removeListener(this)
	}

	override fun onConnected(conn: AClientConnection) {
		if (game.isGameRunning()) {
			conn.getAttribute("color")?.takeIfInstance<Int>()?.let { colorId ->
				val color = ZUser.USER_COLORS[colorId]
				var charsList = game.board.getAllCharacters().filter { it.color == color }
				log.debug("color: $color, chars: ${charsList.joinToString()}")
				if (charsList.isEmpty()) {
					val allInvisible = game.board.getAllCharacters().filter { it.isInvisible }
					val colors = allInvisible.map { it.color }.toSet().toList()
					log.debug("allColors: ${colors.joinToString()}")
					charsList = allInvisible.filter { it.color == colors[0] }
				}
				if (charsList.isNotEmpty()) {
					conn.addListener(this)
					val user: ZUser = ZUserMP(conn)
					clientToUserMap[conn.name] = user
					game.server = server
					charsList.forEach {
						it.isInvisible = false
						it.isReady = true
						user.addCharacter(it)
					}
					game.addUser(user)
					broadcastPlayerStarted(user.name)
					game.setUserColorId(user, colorId)
					conn.sendCommand(newLoadQuest(game.quest.quest))
					val assignments = charsList.map {
						Assignee(it.type, conn.name, colorId, true)
					}
					conn.sendCommand(newInit(colorId, game.quest.quest))
					conn.sendCommand(newUpdateGameCommand(game))
					game.characterRenderer.addMessage(
						conn.displayName + " has joined",
						user.getColor()
					)
					game.readyLock.release()
					return
				}
			}

			conn.disconnect("Game in Progress")
			return
		}

		conn.addListener(this)
		val user: ZUser = ZUserMP(conn)
		val usedColors = game.getUsers().map { it.colorId }
		log.debug("Used colors: ${usedColors.map { ZUser.USER_COLOR_NAMES[it] }.joinToString()}")
		if (usedColors.contains(user.colorId)) {
			val availableColors = Array(ZUser.USER_COLORS.size) { it }.filter {
				!usedColors.contains(it)
			}
			val color = availableColors[0]
			log.debug("User ${user.name} is being reassigned color ${ZUser.USER_COLOR_NAMES[color]}")
			game.setUserColorId(user, color)
		}
		clientToUserMap[conn.name] = user
		conn.sendCommand(newInit(user.colorId, game.quest.quest))
		game.characterRenderer.addMessage(conn.displayName + " has joined", user.getColor())
		conn.sendCommand(newOpenAssignmentsDialog(maxCharacters, playerAssignments.values.toList()))
	}

	override fun onReconnected(conn: AClientConnection) {
		val user = clientToUserMap[conn.name]
		log.debug("onReconnected user: $user")
		if (user != null) {
			game.addUser(user)
			conn.sendCommand(newInit(user.colorId, game.quest.quest))
			game.characterRenderer.addMessage(conn.displayName + " has rejoined", user.getColor())
			user.players.map { game.board.getCharacter(it) }.forEach {
				it.isInvisible = false
			}
			game.boardRenderer.redraw()
		}
	}

	override fun onDisconnected(conn: AClientConnection, reason: String?) {
		// TODO: Put up a dialog waiting for client to reconnect otherwise set their characters to invisible and to stop moving
		val user = clientToUserMap[conn.name]
		if (user != null) {
			user.players.map { game.board.getCharacter(it) }.forEach {
				it.isInvisible = true
			}
			game.removeUser(user)
			game.characterRenderer.addMessage("${conn.displayName} has disconnected because $reason", user.getColor())
			game.boardRenderer.redraw()
		}
	}

	override fun onCommand(conn: AClientConnection, cmd: GameCommand) {
		parseCLCommand(conn, cmd)
	}

	override fun onChooseCharacter(conn: AClientConnection, name: ZPlayerName, checked: Boolean) {
		val a = playerAssignments[name]
		val user = clientToUserMap[conn.name]
		log.debug("onChooseCharacter assignee: $a, user: ${user?.name}")
		if (a != null && user != null) {
			a.checked = checked
			a.isAssingedToMe = false
			if (checked) {
				a.userName = conn.displayName
				a.color = user.colorId
				val c = game.addCharacter(name)
				user.addCharacter(c)
			} else {
				a.color = -1
				a.userName = "??"
				game.removeCharacter(name)?.let {
					user.removeCharacter(it)
				}
			}
			server.broadcastCommand(newAssignPlayer(a))
			playerChooser?.postNotifyUpdateAssignee(a)
			game.boardRenderer.redraw()
		}
	}

	override fun onStartPressed(conn: AClientConnection) {
		clientToUserMap[conn.name]?.let { user ->
			game.addUser(user)
			broadcastPlayerStarted(user.name)
		}
	}

	fun broadcastPlayerStarted(userName: String) {
		server.broadcastCommand(newPlayerStartedCommand(userName, ++numStarted, game.getConnectedUsers().size))
		if (numStarted > server.numConnectedClients) {
			game.server = server
			activity.runOnUiThread {
				activity.startGame()
				broadcastUpdateGame()
			}
		}
	}

	override fun onUndoPressed(conn: AClientConnection) {
		activity.runOnUiThread { game.undo() }
	}

	override fun onColorPickerPressed(conn: AClientConnection) {
		val currentColors = game.getUsers().map { it.colorId }
		val colorOptions =
			Array(ZUser.USER_COLORS.size) { it }.filter { !currentColors.contains(it) }
		conn.sendCommand(newColorOptions(colorOptions))
	}

	override fun onError(e: Exception) {
		log.error(e)
		game.addPlayerComponentMessage("Error:" + e.javaClass.simpleName + ":" + e.message)
	}

	override fun onConnectionStatusChanged(c: AClientConnection, status: ConnectionStatus) {
		val connections = getConnectedUsers().map {
			Triple(
				it.first, it.second,
				ConnectedUser(
					it.second.name,
					it.second.getColor(),
					it.first.isConnected,
					ConnectionStatus.from(it.first.connectionSpeed),
					game.getStartUser().colorId == it.second.colorId
				)
			)
		}
		// each client gets all the players including the server and minus themselves
		connections.forEach { triple ->
			// the server player takes the connection status of the target client
			val connectionsForClient = mutableListOf(
				ConnectedUser(
					game.thisUser.name,
					game.thisUser.getColor(),
					true,
					triple.third.status,
					triple.third.startUser
				)
			).also {
				it.addAll(connections.filter { it != triple.second }.map { it.third })
			}
			triple.first.sendCommand(newConnectionsInfo(connectionsForClient))
		}
	}

	fun broadcastUpdateGame() {
		server.broadcastCommand(newUpdateGameCommand(game))
		//previous.copyFrom(game);
	}

	fun broadcastShowChooser(assignments: List<Assignee>) {
		server.broadcastCommand(newOpenAssignmentsDialog(maxCharacters, assignments))
	}

	companion object {
		var log = LoggerFactory.getLogger(ZServerMgr::class.java)
	}

	init {
		addListener(this)
		server.addListener(this)
		if (!game.isGameRunning()) {
			showChooser()
		}
	}

	fun showChooser() {
		playerAssignments.clear()
		for (c in activity.charLocks) {
			val a = Assignee(c)
			playerAssignments[c.player] = a
		}
//		activity.thisUser.setColor(game.board, nextColor())
		val assignments: List<Assignee> = ArrayList(playerAssignments.values)
		Collections.sort(assignments)
		playerChooser =
			object : CharacterChooserDialogMP(activity, assignments, true, maxCharacters) {
				override fun onAssigneeChecked(assignee: Assignee, checked: Boolean) {
					synchronized(playerAssignments) {
						assignee.checked = checked
						if (checked) {
							assignee.userName = activity.displayName
							assignee.color = activity.thisUser.colorId
							assignee.isAssingedToMe = true
							val c = game.addCharacter(assignee.name)
							activity.thisUser.addCharacter(c)
						} else {
						assignee.userName = "??"
						assignee.color = -1
						assignee.isAssingedToMe = false
						game.removeCharacter(assignee.name)?.let {
							activity.thisUser.removeCharacter(it)
						}
					}
					postNotifyUpdateAssignee(assignee)
					server.broadcastCommand(newAssignPlayer(assignee))
						game.boardRenderer.redraw()
					}
				}

				override fun onStart() {
					dialog.dismiss()
					broadcastPlayerStarted(activity.thisUser.name)
				}
			}
	}

	fun setMaxCharactersPerPlayer(max: Int) {
		server.broadcastCommand(newUpdateMaxCharactersPerPlayer(max))
	}

	fun getConnectedUsers(): List<Pair<AClientConnection, ZUser>> =
		server.connectionValues.filter { clientToUserMap[it.name] != null }.map {
			Pair(it, clientToUserMap[it.name]!!)
		}
}